paperoni/commands/command_collect_semantic_scholar.py

```diff
@@ -0,0 +1,281 @@
+
+    def __init__(self, filename):
+        self.db = Database(filename)
+        self.papers_added = set()
+        self.papers_excluded = set()
+
+    def __contains__(self, paper: Paper):
+        if paper in self.papers_excluded:
+            return False
+        elif paper in self.papers_added:
+            return True
+        else:
+            return self.db.select_id_from_values(
+                "paper", "paper_id", title=paper.title, abstract=paper.abstract
+            )
+
+    def add(self, paper: Paper):
+        if paper in self.papers_excluded:
+            self.papers_excluded.remove(paper)
+        self.papers_added.add(paper)
+
+    def exclude(self, paper: Paper):
+        if paper in self.papers_added:
+            self.papers_added.remove(paper)
+        self.papers_excluded.add(paper)
+
+    def excludes(self, paper: Paper):
+        return paper in self.papers_excluded
+
+    def save(self):
+        # Delete excluded papers.
+        for paper in self.papers_excluded:
+            if paper.abstract is None:
+                self.db.modify(
+                    "DELETE FROM paper WHERE title = ? AND abstract IS NULL",
+                    [paper.title],
+                )
+            else:
+                self.db.modify(
+                    "DELETE FROM paper WHERE title = ? AND abstract = ?",
+                    (paper.title, paper.abstract),
+                )
+        # Add or update papers.
+        for paper in self.papers_added:
+            self._save_paper(paper)
+
+    def _save_paper(self, paper: Paper):
+        db = self.db
+        author_indices = []
+        topic_indices = []
+        # Paper: check MAG ID then title+abstract.
+        paper_id = (
+            db.select_id_from_values(
+                "paper_link",
+                "paper_id",
+                link_type="MAG",
+                link=paper.get_ref("MAG"),
+            )
+            or db.select_id_from_values(
+                "paper", "paper_id", title=paper.title, abstract=paper.abstract
+            )
+            or db.insert(
+                "paper",
+                ("title", "abstract", "citation_count"),
+                (paper.title, paper.abstract, paper.citation_count),
+            )
+        )
+        # Links
+        for link in paper.links:
+            if not db.count(
+                "paper_link",
+                "paper_id",
+                "link_type = ? AND link = ?",
+                (link.type, link.ref),
+            ):
+                db.insert(
+                    "paper_link",
+                    ("paper_id", "link_type", "link"),
+                    (paper_id, link.type, link.ref),
+                )
+        # Authors
+        for author in paper.authors:
+            # Author
+            author_id = db.select_id(
+                "author", "author_id", "author_name = ?", [author.name]
+            ) or db.insert("author", ["author_name"], [author.name])
```

Comment:
The same author can be listed under different names in different papers, e.g. "John Smith" and "J Smith", so it would be better to get the ID from the links, as for the papers.